File: implementations/python/cid.py
import base64
import hashlib
from pathlib import Path
from typing import Tuple
from urllib.request import Request, urlopen


BASE_DIR = Path(__file__).resolve().parents[2]
EXAMPLES_DIR = BASE_DIR / "examples"
CIDS_DIR = BASE_DIR / "cids"


def to_base64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode().rstrip("=")


def encode_length(length: int) -> str:
    return to_base64url(length.to_bytes(6, "big"))


def compute_cid(content: bytes) -> str:
    prefix = encode_length(len(content))
    if len(content) <= 64:
        suffix = to_base64url(content)
    else:
        suffix = to_base64url(hashlib.sha512(content).digest())
    return prefix + suffix


def download_cid(base_url: str, cid: str) -> Tuple[bytes, str, bool]:
    """
    Download content from a URL and validate its CID.
    
    Args:
        base_url: Base URL to download from (e.g., "https://256t.org")
        cid: Expected content identifier
        
    Returns:
        Tuple of (content, computed_cid, is_valid)
        - content: The downloaded content as bytes
        - computed_cid: The CID computed from the downloaded content
        - is_valid: True if computed_cid matches the expected cid
    """
    url = f"{base_url.rstrip('/')}/{cid}"
    req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urlopen(req) as response:
        content = response.read()
    computed = compute_cid(content)
    return content, computed, computed == cid


__all__ = [
    "BASE_DIR",
    "EXAMPLES_DIR",
    "CIDS_DIR",
    "compute_cid",
    "download_cid",
]

File: implementations/ocaml/cid.ml
let to_base64url data =
  Base64.encode_exn ~pad:false ~alphabet:Base64.uri_safe_alphabet data

let encode_length length =
  let bytes = Bytes.make 6 '\x00' in
  for i = 0 to 5 do
    Bytes.set bytes (5 - i) (Char.chr ((length lsr (8 * i)) land 0xFF))
  done;
  bytes |> Bytes.to_string |> to_base64url

let compute_cid content =
  let length = String.length content in
  let prefix = encode_length length in
  let suffix =
    if length <= 64 then
      to_base64url content
    else
      Digestif.SHA512.digest_string content
      |> Digestif.SHA512.to_raw_string
      |> to_base64url
  in
  prefix ^ suffix

let executable_dir () =
  let exe = Sys.executable_name in
  let exe = if Filename.is_relative exe then Filename.concat (Sys.getcwd ()) exe else exe in
  Filename.dirname exe

let base_dir () = executable_dir () |> Filename.dirname |> Filename.dirname
let examples_dir () = Filename.concat (base_dir ()) "examples"
let cids_dir () = Filename.concat (base_dir ()) "cids"
